project/src/com/google/daggerquery/server/Server.java

```diff
@@ -0,0 +1,73 @@
+package com.google.daggerquery.server;
+
+import com.google.common.collect.ImmutableList;
+import com.google.daggerquery.executor.QueryExecutor;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.server.RoutingHandler;
+import io.undertow.Undertow;
+import io.undertow.util.HttpString;
+import io.undertow.util.StatusCodes;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Deque;
+import java.util.stream.Collectors;
+
+public class Server {
+
+  private static final int port = 4921;
+  private static final String host = "localhost";
+
+  /**
+   * Starts a server that accepts a single <b>GET</b> request with the specified query.
+   */
+  public static void main(String[] args) {
+    RoutingHandler routingHandler = new RoutingHandler();
+    routingHandler.add("GET", "/daggerquery/{query}", Server::executeQuery);
+
+    Undertow undertowServer = Undertow.builder()
+        .addHttpListener(port, host)
+        .setHandler(routingHandler)
+        .build();
+    undertowServer.start();
+  }
+
+  /**
+   * Extracts the query from {@link HttpServerExchange} and tries to execute it.
+   *
+   * <p>If http response didn't contain <b>query</b> parameter then this method does nothing.
+   *
+   * <p>Otherwise it executes query and if it was successful sends correct subgraphs from the
+   * origin binding graphs as the result.
+   *
+   * <p>For <b>deps</b> query returns a list with a string representation of source node's dependencies.
+   * For <b>allpaths</b> and <b>somepath</b> returns a formatted {@link com.google.daggerquery.executor.models.Query.Path}
+   * representation with the node names separated by an arrow.
+   */
+  private static void executeQuery(HttpServerExchange exchange) {
+    exchange.getResponseHeaders().put(new HttpString("Access-Control-Allow-Origin"), "*");
+    exchange.getResponseHeaders().put(new HttpString("Access-Control-Allow-Methods"), "GET");
+
+    Deque<String> deque = exchange.getQueryParameters().get("query");
+    if (deque == null) {
+      return;
+    }
+
+    String[] args = deque.stream().filter(value -> value.length() > 0).toArray(String[]::new);
```

Comment:
I am not sure about this line: I mentioned that `getQueryParameters().get("query")` returns a deque which always contains one more element - just an empty string. My request looks like this: `http://localhost:4921/daggerquery/?query=deps&query=com.google.Beach`.

It is not documented behavior. In this line, I am filtering all empty strings from the array before passing them to the `QueryExecutor`.

Reply:
I'm not sure if the order of repeated params are guaranteed to be preserved, if not this design could be fragile, say the URL ends up as `query=com.google.Beach&query=deps`. How about taking the query as one single argument, e.g. `query=deps%20com.google.Beach`?

Reply:
O, great solution, thanks. It also will help us to get rid of these filtering calls because we can just take the first element from a deque.